fix #277: randomEmail hexEncodedBuffer contains null chars

edarchis opened the pull request:
When the `EmailTransformer.hexEncodedRandomBytesBuf` is initialized, it is allocated as a 64 bytes buffer (`emailTransformerGeneratorSize`). The hex encoded buffer has to be at least twice the size of the content to encode. If it's larger, it crashes (this might also be an issue but a more obvious one). If it's smaller, it'll contain nulls.

To work around the issue, when not using a template for the local part of the email, we can just take twice the content size as buffer.

wwoytenko commented:
Hi! Thank you for your contribution. Could you cover it with unit test as well?

Let me know if any help is needed

edarchis commented:
> Hi! Thank you for your contribution. Could you cover it with unit test as well?
> 
> Let me know if any help is needed

I would love to but didn't find existing unit tests for this transformer. As I have done any golang in years and am still discovering the tool, I didn't feel like starting one. If you have some pointer to other similar unit tests that I could take inspiration from, I would gladly add the email tests.

Looking back at this bug, it went under the radar because if you don't specify a length, it'll default to the exact size that will work. The reason I tripped on it is because I copy-pasted the sample from the docs that specifies the length... 🤷

edarchis commented:
Hi @wwoytenko, this should be good to go now. It was worth the effort as it showed that `{{.random_string}}` had the same issue. Please check the message of the last commit about the size parameter.

To show that the test did trigger the original bug:
![SCR-20250320-ojcb](https://github.com/user-attachments/assets/6ecff7b0-e8bb-427b-8891-def75537dd0a)

wwoytenko commented on internal/db/postgres/transformers/email.go:
```diff
@@ -184,33 +184,37 @@ func NewEmailTransformer(ctx context.Context, driver *toolkit.Driver, parameters
 		return nil, nil, fmt.Errorf("unable to get generator: %w", err)
 	}
+	randomBufferSize := emailTransformerGeneratorSize
+	if engine == "random" {
```
I think it should be working this way for `engine=hash` as well. Probably you can remove variable `randomBufferSize` and the if condition and just use `maxRandomLength` directly as buf length. 

A const `emailTransformerGeneratorSize` can be removed.

It would be like this

```
	g, err := getGenerateEngine(ctx, engine, maxRandomLength)
	if err != nil {
		return nil, nil, fmt.Errorf("unable to get generator: %w", err)
	}

	return &EmailTransformer{
		g:                        g,
		columnName:               columnName,
		keepNull:                 keepNull,
		affectedColumns:          affectedColumns,
		columnIdx:                idx,
		templateCtx:              make(map[string]any, 10),
		keepOriginalDomain:       keepOriginalDomain,
		domains:                  domains,
		localPartTemplate:        localTmpl,
		domainTemplate:           domainTmpl,
		validate:                 validate,
		buf:                      bytes.NewBuffer(nil),
		hexEncodedRandomBytesBuf: make([]byte, hex.EncodedLen(maxRandomLength)),
		rctx:                     rrctx,
	}, nil, nil
```

wwoytenko commented on internal/db/postgres/transformers/email_test.go:
```diff
@@ -0,0 +1,321 @@
+		expectedErr string
+	}{
+		{
+			name:       "common",
```
I suggest to add test that covers `engine: hash`. For example now it fails. You can just copy and paste existing test for random and replace to `hash`.

For example

```
		{
			name:       "common hash",
			original:   "dupont@mycompany.com",
			columnName: "data",
			params: map[string]toolkit.ParamsValue{
				"column": toolkit.ParamsValue("data"),
				"engine": toolkit.ParamsValue("hash"),
			},
			validateFn: func(t *testing.T, originalEmail, transformedEmail string) {
				assert.True(t, validEmailRegexp.MatchString(transformedEmail))
				assert.NotEqual(t, originalEmail, transformedEmail)
			},
		},
```

wwoytenko reviewed (changes requested):
Thank you so much for your detailed investigation!

Just a few simple things to resolve, and I believe we can proceed with the merge.

edarchis commented:
You're perfectly right of course. I implemented your suggestions and added a couple more tests for the validate parameter. That one was working fine luckily.

wwoytenko reviewed (approved):
LGTM
